tests/test_cli.py

```diff
@@ -32,16 +32,17 @@ def _get_subprocess_env():
 
 
 SUBPROCESS_ENV = _get_subprocess_env()
+call = partial(subprocess.call, env=SUBPROCESS_ENV, shell=True)
```

Comment:
Why is `shell=True` needed?

Reply:
I'll be honest I don't completely understand the underlying mechanics, however what I've found empirically is that it's the only way to get the return code `0` back successfully while keeping both Linux and Windows happy. Without it, we get the error in the issue description. [The Python docs](https://docs.python.org/3/library/subprocess.html#frequently-used-arguments) and StackOverflow suggest that the behaviour is highly platform dependent, a good summary of using string vs. list and shell=True vs. shell=False is here: https://stackoverflow.com/a/15109975

Reply:
we could use `shlex.split` as well...

Reply:
**TL;DR: Let's change the conditional?**

I've tested each combination of using string vs list and shell vs. no shell and this was the only way to successfully get a return code of `0` when hg is present and a nonzero return code when hg is not present.

The general advice I've seen is to avoid using the list style when using `shell=True`, here's a Python bug report showing it's a commonly confused topic: https://bugs.python.org/issue21347

> It is best to always use a string if `shell=True` is necessary.
> It is a common confusion #20344, msg98732, #7839
> http://stackoverflow.com/questions/21029154/understanding-python-subprocess-check-outputs-first-argument-and-shell-true
> http://stackoverflow.com/questions/20787712/start-openoffice-process-with-python-to-use-with-pyuno-using-subprocess
> http://stackoverflow.com/questions/17880847/python-subprocess-error-in-using-cp
> http://stackoverflow.com/questions/17226912/why-does-simple-echo-in-subprocess-not-working
> http://stackoverflow.com/questions/15109665/subprocess-call-using-string-vs-using-list
> http://stackoverflow.com/questions/20128114/pythons-subprocess-does-not-interpret-as-expected-on-cygwin
> http://stackoverflow.com/questions/16840427/python-on-linux-subprocess-popen-works-weird-with-shell-true

That being said though, I had a think about it overnight and if we can't make everyone happy with the `shell=True`, we could rethink the conditional check and avoid this confusion altogether by using another utility? Are you guys open to adding a new third party requirement to the `[testenv]` deps in tox? If so, I think this project would simplify the check: https://pypi.org/project/whichcraft/
e.g.
```
from whichcraft import which
...
xfail_if_no_hg = pytest.mark.xfail(
  which("hg") is None
  reason="hg is not installed"
)
```

If third party packages for the test suite are to be avoided, then we could implement a thin but platform dependant `which` / `where` check:
e.g.
```
xfail_if_no_hg = pytest.mark.xfail(
  (call(["where", "hg"]) != 0) if sys.platform == "win32" else (call(["which", "hg"]) != 0),
  reason="hg is not installed"
)
```

I've learnt a lot here, happy to go with whatever's easiest for you guys!

Reply:
I've also found that not using `shell=True` on Windows raises a FileNotFoundError because `hg.exe` can not be found on disk .  Setting `shell=True` fixes that -- it starts the *shell* and passes 'hg' as the first argument.

I would not use another third-party package and using `which` and `where` seems risky as well -- do we know for sure that this is reliable in all virtual envs and CI VMs?  What if `git` is not an executable but a shell alias for another tool which is git compatible?

It looks like `shell=True` is the way to go, in combination with using a string instead of a list.

Reply:
If `shell=True` works with a single string then that should be the way to go. There's no variables and it's only test code so there's no real security risk.